core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,214 @@
+          responseQueue.add(GetChildrenResponse(rc, path, ctx, Option(children).map(_.asScala).orNull, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+      case ExistsWithWatcherRequest(path) => zooKeeper.exists(path, true, new StatCallback {
+        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
+          responseQueue.add(ExistsResponse(rc, path, ctx, stat))
+          countDownLatch.countDown()
+        }}, null)
+      case GetChildrenWithWatcherRequest(path) => zooKeeper.getChildren(path, true, new Children2Callback {
+        override def processResult(rc: Int, path: String, ctx: Any, children: java.util.List[String], stat: Stat) = {
+          responseQueue.add(GetChildrenResponse(rc, path, ctx, Option(children).map(_.asScala).orNull, stat))
+          countDownLatch.countDown()
+        }}, null)
+    }
+    countDownLatch.await()
+    responseQueue.asScala.toSeq
+  }
+
+  def waitUntilConnectedOrExpired: Boolean = inLock(isConnectedOrExpiredLock) {
+    var state = zooKeeper.getState
+    while (!state.isConnected && state.isAlive) {
+      isConnectedOrExpiredCondition.await()
+      state = zooKeeper.getState
+    }
+    state.isConnected
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): ExistsResponse = inReadLock(initializationLock) {
+    registerZNodeChangeHandlers(Seq(zNodeChangeHandler)).head
+  }
+
+  def registerZNodeChangeHandlers(handlers: Seq[ZNodeChangeHandler]): Seq[ExistsResponse] = inReadLock(initializationLock) {
+    handlers.foreach(handler => zNodeChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => ExistsWithWatcherRequest(handler.path))
+    handle(asyncRequests).asInstanceOf[Seq[ExistsResponse]]
+  }
+
+  def unregisterZNodeChangeHandler(path: String): Unit = inReadLock(initializationLock) {
+    zNodeChangeHandlers.remove(path)
+  }
+
+  def registerZNodeChildChangeHandler(zNodeChildChangeHandler: ZNodeChildChangeHandler): GetChildrenResponse = inReadLock(initializationLock) {
+    registerZNodeChildChangeHandlers(Seq(zNodeChildChangeHandler)).head
+  }
+
+  def registerZNodeChildChangeHandlers(handlers: Seq[ZNodeChildChangeHandler]): Seq[GetChildrenResponse] = inReadLock(initializationLock) {
+    handlers.foreach(handler => zNodeChildChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => GetChildrenWithWatcherRequest(handler.path))
+    handle(asyncRequests).asInstanceOf[Seq[GetChildrenResponse]]
+  }
+
+  def unregisterZNodeChildChangeHandler(path: String): Unit = inReadLock(initializationLock) {
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def close(): Unit = inWriteLock(initializationLock) {
+    zNodeChangeHandlers.clear()
+    zNodeChildChangeHandlers.clear()
+    zooKeeper.close()
+  }
+
+  private def initialize(): Unit = {
+    if (!zooKeeper.getState.isAlive) {
+      zNodeChangeHandlers.clear()
+      zNodeChildChangeHandlers.clear()
```

Comment:
It would be useful to add some info level logging to indicate that a new ZK session is created.